core/cloudflow-operator-actions/src/main/scala/cloudflow/operator/action/Action.scala

```diff
@@ -208,195 +199,243 @@ abstract class SingleResourceAction[T <: ObjectResource] extends ResourceAction[
+          val prevMetadata = existingResource.getMetadata
+          val newMetadata  = resource.getMetadata
+
+          newMetadata.setResourceVersion(prevMetadata.getResourceVersion)
           recoverFromError(
-            client.update(resourceVersionUpdated),
+            Future { client.resource(resource).createOrReplace() },
             client,
             nextRetries,
             executeCreate
           )
         }
-        .getOrElse(recoverFromError(client.create(resource), client, nextRetries, executeCreate))
+        .getOrElse(
+          recoverFromError(Future { client.resource(resource).createOrReplace() }, client, nextRetries, executeCreate)
+        )
     } yield res
   }
 }
 
 /**
  * Captures the update of the resource.
  */
-class CreateOrPatchAction[T <: ObjectResource, O <: Patch](
+class CreateOrPatchAction[T <: HasMetadata](
     val resource: T,
-    val patch: O,
-    implicit val format: Format[T],
-    implicit val patchWriter: Writes[O],
-    implicit val resourceDefinition: ResourceDefinition[T]
-) extends SingleResourceAction[T] {
+    val patch: T
+)(implicit ct: ClassTag[T])
+    extends SingleResourceAction[T] {
 
   val name = "create-or-patch"
 
   /**
    * Updates the resource, without changing the `resourceVersion`.
    */
-  def execute(client: KubernetesClient)(implicit sys: ActorSystem, ec: ExecutionContext, lc: LoggingContext): Future[Action] =
+  def execute(client: KubernetesClient)(implicit sys: ActorSystem, ec: ExecutionContext): Future[Action] =
     for {
       result <- executeCreateOrPatch(client)
-    } yield new CreateOrPatchAction(result, patch, format, patchWriter, resourceDefinition)
+    } yield new CreateOrPatchAction(result, patch)
 
   private def executeCreateOrPatch(
       client: KubernetesClient,
       retries: Int = 60
-  )(implicit sys: ActorSystem, ec: ExecutionContext, lc: LoggingContext): Future[T] = {
+  )(implicit sys: ActorSystem, ec: ExecutionContext): Future[T] = {
     val nextRetries = retries - 1
 
+    val existing = Try {
+      client
+        .resource(resource)
+        .inNamespace(namespace.getOrElse(client.getNamespace))
+        .get()
+    }.toOption
     for {
-      existing <- client
-        .usingNamespace(namespace.getOrElse(client.namespaceName))
-        .getOption[T](resource.name)
       res <- existing
-        .map(_ => recoverFromError(client.patch(resource.name, patch, Some(resource.ns)), client, nextRetries, executeCreateOrPatch))
-        .getOrElse(recoverFromError(client.create(resource), client, nextRetries, executeCreateOrPatch))
+        .map(_ =>
+          recoverFromError(
+            Future {
+              client.secrets().inNamespace().withName()
+              client
+                .resource(resource)
+                .inNamespace(resource.getMetadata.getNamespace)
+                .get()
+                .asInstanceOf[Resource[T, Doneable]] // TODO: this is not going to work ...
```

Comment:
@RayRoestenburg fyi:
This is still a problem, probably `Resource[T, Doneable]` should pop up somewhere in the signature.